Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,62 @@
+require "delegate"
+require "English"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
```

Comment:
> [This] class does more than just the selection; it actually performs the I/O, too, to some extent.

Very good point. :+1:

The class was initially meant to be no more than a wrapper for `IO::select`. One of the class-level comments I just have added says:

> The class `IOSelector` is a wrapper for `IO::select` with the added benefit that it spans the streams' lifetimes.

Apart from `IO::select`, the I/O that happens would be `eof?` and `close_read`. My idea of the class was that the client _remains fully responsible for reading._ (Otherwise we could have simply used `Open3::capture3`.)

Somehow, in #2466, I allowed a convenience method `#binread_nonblock` to creep in, which _does_ include reading, just to spare the clients from writing a few more lines. Maybe that was a bad idea; It’s definitely beyond what a _selector_ should do.

It feels like we’re having a hard time to find a good name for the class. It also doesn’t seem immediately obvious what the class does. ([@MikeMcQuaid has mentioned it a few days ago.](https://github.com/Homebrew/brew/pull/2762#issuecomment-307571782)) This raises a few red flags for me. I sense a design smell here, not a naming issue. The class wants to do too much.

I figure it would help if I removed all the non-essential I/O code (i. e. the `#binread_nonblock` convenience method and the `close_streams` private method), and let the client handle that for now. That would leave the `IOSelector` class responsible for `IO::select` and `eof?`, and nothing else.

Thoughts?